pkg/agent/externalnode/external_node_controller_linux_test.go

```diff
@@ -0,0 +1,562 @@
+				).Times(1)
+			}
+			err := c.updateExternalNode(tt.preExternalNode, tt.curExternalNode)
+			assert.NoError(t, err)
+			assert.Equal(t, tt.syncedExternalNode, c.syncedExternalNode)
+		})
+	}
+}
+
+func TestAddInterface(t *testing.T) {
+	_, cidr, _ := net.ParseCIDR("172.16.1.0/24")
+
+	for _, tt := range []struct {
+		name                    string
+		iface                   *interfacestore.InterfaceConfig
+		ips                     []string
+		preIPs                  []string
+		portData                *ovsconfig.OVSPortData
+		preEEName               string
+		uplinkPortUUID          string
+		ifaceExist              bool
+		getInterfaceConfigParam mockGetInterfaceConfigParam
+		ifAddInterface          bool
+	}{
+		{
+			name:           "add interface successfully",
+			iface:          &intf1,
+			ifaceExist:     true,
+			ips:            []string{"10.20.30.40"},
+			preIPs:         []string{"10.20.30.50"},
+			preEEName:      "externalEntity",
+			ifAddInterface: true,
+		},
+		{
+			name:           "skip adding interface",
+			iface:          &intf1,
+			ifaceExist:     true,
+			ips:            []string{"10.20.30.40"},
+			preIPs:         []string{"10.20.30.40"},
+			preEEName:      "eeName",
+			ifAddInterface: false,
+		},
+		{
+			name:       "add new interface",
+			iface:      &intf1,
+			ifaceExist: false,
+			ips:        []string{"10.20.30.40"},
+			preIPs:     []string{"10.20.30.40"},
+			preEEName:  "externalEntity",
+			getInterfaceConfigParam: mockGetInterfaceConfigParam{
+				&net.Interface{
+					Index:        1,
+					HardwareAddr: net.HardwareAddr{0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88},
+					MTU:          1500,
+				},
+				[]*net.IPNet{cidr},
+				nil,
+			},
+			ifAddInterface: true,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			c := newExternalNodeController(t)
+			mockIfaceStore.EXPECT().GetInterfaceByName(tt.iface.InterfaceName).Return(tt.iface, tt.ifaceExist).Times(1)
+			ips := make([]net.IP, 0, len(tt.ips))
+			for _, ip := range tt.ips {
+				ips = append(ips, net.ParseIP(ip))
+			}
+			iface := &interfacestore.InterfaceConfig{
+				InterfaceName: tt.iface.InterfaceName,
+				Type:          interfacestore.ExternalEntityInterface,
+				IPs:           ips,
+				OVSPortConfig: tt.iface.OVSPortConfig,
+				EntityInterfaceConfig: &interfacestore.EntityInterfaceConfig{
+					EntityName:      eeName,
+					EntityNamespace: tt.iface.EntityNamespace,
+					UplinkPort:      tt.iface.UplinkPort,
+				},
+			}
+			if !tt.ifaceExist {
+				iface.EntityInterfaceConfig.EntityNamespace = eeNamespace
+			}
+			addInterfaceTimes := 0
+			if tt.ifAddInterface {
+				addInterfaceTimes = 1
+			}
+			mockIfaceStore.EXPECT().AddInterface(iface).Times(addInterfaceTimes)
+			if !tt.ifaceExist {
+				defer mockGetInterfaceConfig([]mockGetInterfaceConfigParam{tt.getInterfaceConfigParam})()
+				defer mockConfigureLinkRoutes(nil)()
+				defer mockConfigureLinkAddresses(nil)()
+				defer mockRemoveLinkRoutes(nil)()
```

Comment:
same comment as above: move the expected output to the table, otherwise it's hard to figure out what's the expectation from the various branches of the test code.

Reply:
updated